python/SimpleBalancer.py

```diff
@@ -714,7 +744,78 @@ def getEstLoad(self,group,targetPrefix):
       else:
         return percentTotal
 
-  
+  # finds two prefixes that are next to each other
+  # when merged, bw is less than configurable threshold
+  # deletes the two prefixes and adds the candidate prefix
+  def merge(self):
+      self.logger.debug("Balance By Merge")
+      subnetDict = self.mergeContiguousPrefixes(self.prefix_list)
+      
+      if not subnetDict:
+          return
+
+      for candidatePrefix in subnetDict:
+          prefix_list = []
+          prefix_list = subnetDict[candidatePrefix]
+          
+          if len(prefix_list) != 2: continue
+          
+          prefix_a = prefix_list[0]
+          prefix_b = prefix_list[1]
+      
+          aggBW = (self.prefixBW[prefix_a]/1000/1000) + (self.prefixBW[prefix_b]/1000/1000)
+ 
+          self.logger.error("Prefixes :"  + str(prefix_a) + ", " + str(prefix_b) + " Aggregate BW : " + str(aggBW) + " Mbps" )
+          self.logger.error("Configurable Threshold :" + str(self.sensorConfigurableThreshold))
+
+          #if the bw is less than configurable threshold.
+          if aggBW < self.sensorConfigurableThreshold:
+              minLoad         = 100
+              minSensor       = ""
+
+              for group in self.groups.keys():
+
+                  if(not self.getGroupStatus(group)): continue
+
+              load = self.getGroupLoad(group)
+
+              if(load < minLoad):
+                  minLoad = load
+                  minSensor = group;
+
+              self.logger.debug("Min Group: " + str(minSensor))
+              group_a = self.getPrefixGroup(prefix_a)
+              group_b = self.getPrefixGroup(prefix_b)
+
+              if candidatePrefix.version == 4:
+                  if(candidatePrefix.prefixlen > self.mostSpecificPrefixLen):
+                      return
+              elif candidatePrefix.version == 6:
+                  if(candidatePrefix.prefixlen > self.ipv6MostSpecificPrefixLen):
+                      return
+              else:
+                  return
+
+              #delete the prefixes
+              self.logger.error("Merging Prefixes " + str(prefix_a) + ", " + str(prefix_b))
+              
+              self.delGroupPrefix(group_a,prefix_a)
+              self.delGroupPrefix(group_b,prefix_b)
+              
+              #add the candidate prefix to min sesnor with aggBW
+              self.logger.info("Adding Prefix : " + str(candidatePrefix) + " to " + str(minSensor) + " with bw " + str(aggBW) + "Mbps")
+              try:
+                  aggBW = aggBW*1000*1000
+                  self.addGroupPrefix(minSensor, candidatePrefix, aggBW)
+              except DuplicatePrefixError:
+                  self.logger.debug("Already have prefix: " + str(prefix))
+                  return
+              except MaxFlowCountError:
+                  self.logger.debug("Max Flow Count Error")
+                  return
+              return
+
```

Comment:
seems like if we hit an exception we need to add the group prefixes back, otherwise we might loose prefixes we were balancing on